crates/virtual-raindex/src/events/orderbook.rs

```diff
@@ -0,0 +1,545 @@
+    RemoveOrder(&'a RemoveOrderV3),
+    /// Tokens were deposited into a vault.
+    Deposit {
+        event: &'a DepositV2,
+        /// Token decimals required to reconstruct the original float amount.
+        decimals: Option<u8>,
+    },
+    /// Tokens were withdrawn from a vault.
+    Withdraw(&'a WithdrawV2),
+    /// An order was taken against the orderbook.
+    TakeOrder(&'a TakeOrderV3),
+    /// Two orders were cleared against each other.
+    Clear {
+        clear: &'a ClearV3,
+        state_change: &'a AfterClearV2,
+    },
+}
+
+/// Converts a single [`OrderBookEvent`] into one or more [`RaindexMutation`]s.
+///
+/// The deposit variant requires token decimals to rebuild the float value that
+/// the virtual engine tracks. If decimals are missing the function returns
+/// [`RaindexError::TokenDecimalMissing`].
+pub fn orderbook_event_to_mutations(event: OrderBookEvent<'_>) -> Result<Vec<RaindexMutation>> {
+    match event {
+        OrderBookEvent::AddOrder(event) => Ok(vec![RaindexMutation::SetOrders {
+            orders: vec![event.order.clone()],
+        }]),
+        OrderBookEvent::RemoveOrder(event) => Ok(vec![RaindexMutation::RemoveOrders {
+            order_hashes: vec![event.orderHash],
+        }]),
+        OrderBookEvent::Deposit { event, decimals } => {
+            let decimals =
+                decimals.ok_or(RaindexError::TokenDecimalMissing { token: event.token })?;
+
+            let delta = Float::from_fixed_decimal(event.depositAmountUint256, decimals)?;
+            let mutation = RaindexMutation::VaultDeltas {
+                deltas: vec![VaultDelta {
+                    owner: event.sender,
+                    token: event.token,
+                    vault_id: event.vaultId,
+                    delta,
+                }],
+            };
+            Ok(vec![mutation])
+        }
+        OrderBookEvent::Withdraw(event) => {
+            let amount = Float::from_raw(event.withdrawAmount);
+            let delta = (-amount)?;
+            let mutation = RaindexMutation::VaultDeltas {
+                deltas: vec![VaultDelta {
+                    owner: event.sender,
+                    token: event.token,
+                    vault_id: event.vaultId,
+                    delta,
+                }],
+            };
+            Ok(vec![mutation])
+        }
+        OrderBookEvent::TakeOrder(event) => take_order_to_mutations(event),
+        OrderBookEvent::Clear {
+            clear,
+            state_change,
+        } => clear_events_to_mutations(clear, state_change),
+    }
+}
+
+/// Converts an iterator of [`OrderBookEvent`]s into a flat list of mutations.
+pub fn orderbook_events_to_mutations<'a>(
+    events: impl IntoIterator<Item = OrderBookEvent<'a>>,
+) -> Result<Vec<RaindexMutation>> {
+    let mut mutations = Vec::new();
+    for event in events {
+        mutations.extend(orderbook_event_to_mutations(event)?);
+    }
+    Ok(mutations)
+}
```

Comment:
nitpick: consider (.flatten_ok requires itertools)

```suggestion
/// Converts an iterator of [`OrderBookEvent`]s into a flat list of mutations.
pub fn orderbook_events_to_mutations<'a>(
    events: impl IntoIterator<Item = OrderBookEvent<'a>>,
) -> Result<Vec<RaindexMutation>> {
    events
        .into_iter()
        .map(orderbook_event_to_mutations)
        .flatten_ok()
        .collect()
}
```